### PR TITLE
feat: access control

### PR DESCRIPTION
**Pull request by steve24grd**
All Canister update methods now are authenticating callers. However for ease of development, if Network is set to Local, the methods would always authenticate to true, whereas for Mainnet, authenticate only accepts requests from pre-configured custodian-principal set during init. This will allow testing via Candid UI during development phase, when setting Network to Local.

### CODE REVIEW

File: src/icp_prototype_backend/src/lib.rs [hunk 1]
```diff
@@ -62,6 +63,33 @@ impl ToU64Hash for AccountIdentifier {
     }
 }
 
+fn get_network() -> Network { 
+    NETWORK.with(|net| {
+        *net.borrow()
+    })
+}
+
+fn authenticate() -> Result<(), String> {
```
Comment (garudatheo) [inline src/icp_prototype_backend/src/lib.rs:72]: Consider implementing this: https://github.com/AlphaCQ/IC_Utils/blob/main/src/guard.rs

It's pretty interesting.

**Review by garudatheo — CHANGES_REQUESTED** [review]

File: src/icp_prototype_backend/src/lib.rs [hunk 2]
```diff
@@ -25,13 +25,14 @@ use memory::{
...
 use types::{
-    CanisterApiManager, CanisterApiManagerTrait, IcCdkSpawnManager, IcCdkSpawnManagerTrait,
+    Network, CanisterApiManager, CanisterApiManagerTrait, IcCdkSpawnManager, IcCdkSpawnManagerTrait,
     InterCanisterCallManager, InterCanisterCallManagerTrait, Operation, QueryBlocksRequest,
     QueryBlocksResponse, StoredPrincipal, StoredTransactions, SweepStatus, TimerManager,
     TimerManagerTrait, Timestamp,
 };
 
 thread_local! {
+    static NETWORK: RefCell<Network> = RefCell::new(Network::Local);
```
Comment (garudatheo) [inline src/icp_prototype_backend/src/lib.rs:35]: Add this in `memory.rs` to make it part of the permanent memory block.

**Review by garudatheo — COMMENTED** [review]
Don't forget to run `cargo fmt` to implement code formatting.

Also, make sure you test if `cargo test`  is running all the tests.

File: src/icp_prototype_backend/src/lib.rs (hunk 2 above)
Comment (steve24grd) [reply src/icp_prototype_backend/src/lib.rs:35]: added NETWORK as CONNECTED_NETWORK in memory.rs

File: src/icp_prototype_backend/src/lib.rs (hunk 1 above)
Comment (steve24grd) [reply src/icp_prototype_backend/src/lib.rs:73]: implemented and tested guards

**Review by garudatheo — APPROVED** [review]

File: src/icp_prototype_backend/src/types.rs [hunk 3]
```diff
@@ -6,6 +6,74 @@ use ic_ledger_types::{BlockIndex, TransferArgs};
...
+
+pub struct State {
+    pending_requests: BTreeSet<Principal>,
+}
+
+thread_local! {
+    pub static STATE: RefCell<State> = RefCell::new(State{pending_requests: BTreeSet::new()});
+}
+
+#[derive(Deserialize, CandidType, Clone)]
```
Comment (garudatheo) [inline src/icp_prototype_backend/src/types.rs:25]: Add comments to credit the `IC_Utils` library for `CallerGuard`.

**Review by garudatheo — CHANGES_REQUESTED** [review]

**Review by garudatheo — APPROVED** [review]

**Pull request merged**